### PR TITLE
[pkg/ottl] Add the ottl.ParserCollection utility

### PR DESCRIPTION
**Pull request by edmocosta**
#### Description

This PR is part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017, and adds the parser collection utility into the `ottl` package. 

This utility is based on the existing `transformprocessor`'s `ParserCollection` concept and is used to group context's parsers, abstracting the logic to infer the context from the statements, and/or rewriting them adding their paths's context if necessary.

In summary, the general idea is: given the configured contexts, parsers and statements converters, pick the right parser, and transform the parsed result (`ottl.Statements[K]`) into something common to the parser collection (`[R]`). For more details, please have a look at the transformprocessor` refactoring on the [draft](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35050/files#diff-61f39ebf5d942d1b77831a8d13832f92fbf7b138abe08e52259c316f2a679793) using this utility.

Given the `ottl.Parser[K]`'s have different type restrictions/generics (e.g. `TransformContext`), this utility heavily rely on the Go's reflection api, being the exposed interface type-safe, but manipulating untyped objects under the hood.

Considering this code is only executed during the bootstrap, IMO, it should be fine using reflection. I couldn't think about any other solution that would make this utility as generic as it's without reflection. Another approach would be hard-code all the existing ottl contexts's parsers on the collection, having fixed `WithParserX` functions per context and generating a direct dependency on the `ottl/contexts` package.

#### Link to tracking issue
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017

#### Testing
Unit tests

#### Documentation
No changes

### CODE REVIEW

File: pkg/ottl/parser_collection.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+} = (*Parser[any])(nil)
+
+var _ ParsedStatementConverter[any, StatementsGetter, any] = func(
+	_ *ParserCollection[StatementsGetter, any],
+	_ *Parser[any],
+	_ string,
+	_ StatementsGetter,
+	_ []*Statement[any]) (any, error) {
+	return nil, nil
+}
```
Comment (TylerHelmuth) [inline pkg/ottl/parser_collection.go:25]: Do we need this?

File: pkg/ottl/parser_collection.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+	"fmt"
+	"reflect"
+
+	"go.opentelemetry.io/collector/component"
+	"go.uber.org/zap"
+)
+
+var _ interface {
+	ParseStatements(statements []string) ([]*Statement[any], error)
+} = (*Parser[any])(nil)
```
Comment (TylerHelmuth) [inline pkg/ottl/parser_collection.go:16]: Is this enforcing that we'll be able to make the reflective call to `ParseStatements`?

File: pkg/ottl/parser_collection.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+	// GetStatements retrieves the OTTL statements to be parsed
+	GetStatements() []string
+}
+
+// ottlParserWrapper wraps an ottl.Parser using reflection, so it can invoke exported
+// methods without knowing its generic type (transform context).
+type ottlParserWrapper[S StatementsGetter] struct {
+	parser                         reflect.Value
+	prependContextToStatementPaths func(context string, statement string) (string, error)
+}
```
Comment (TylerHelmuth) [inline pkg/ottl/parser_collection.go:38]: Does this type need a generic? I don't see where it uses it.

File: pkg/ottl/parser_collection.go [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+// parserCollectionParser holds an ottlParserWrapper and its respectively
+// statementsConverter function.
+type parserCollectionParser[S StatementsGetter] struct {
+	ottlParser          *ottlParserWrapper[S]
+	statementsConverter statementsConverterWrapper[S]
+}
+
+// ParserCollection is a configurable set of ottl.Parser that can handle multiple OTTL contexts
+// parsings, inferring the context and choosing the right parser for the given statements.
+type ParserCollection[S StatementsGetter, R any] struct {
```
Comment (TylerHelmuth) [inline pkg/ottl/parser_collection.go:110]: Is `S StatementsGetter` required? Would specifying parameters of type `StatementGetter` sufficient? Or even better can we get away with `[]string`?

File: pkg/ottl/parser_collection.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+// ParseStatements parses the given statements into [R] using the configured context's ottl.Parser
+// and subsequently calling the ParsedStatementConverter function.
+// The statement's context is automatically inferred from the [Path.Context] values, choosing the
+// highest priority context found.
+// If no contexts are present in the statements, or if the inferred value is not supported by
+// the [ParserCollection], it returns an error.
+// If parsing the statements fails, it returns the underline [ottl.Parser.ParseStatements] error.
+func (pc *ParserCollection[S, R]) ParseStatements(statements S) (R, error) {
```
Comment (TylerHelmuth) [inline pkg/ottl/parser_collection.go:210]: Instead of taking an interface can this take `[]string` like the parser's ParseStatements?

File: pkg/ottl/parser_collection.go [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+// ParseStatementsWithContext parses the given statements into [R] using the configured
+// context's ottl.Parser and subsequently calling the ParsedStatementConverter function.
+// Differently from ParseStatements, it uses the provided context and does not infer it
+// automatically. The context valuer must be supported by the [ParserCollection],
+// otherwise an error is returned.
+// If the statement's Path does not provide their Path.Context value, the prependPathsContext
+// argument should be set to true, so it rewrites the statements prepending the missing paths
+// contexts.
+// If parsing the statements fails, it returns the underline [ottl.Parser.ParseStatements] error.
+func (pc *ParserCollection[S, R]) ParseStatementsWithContext(context string, statements S, prependPathsContext bool) (R, error) {
```
Comment (TylerHelmuth) [inline pkg/ottl/parser_collection.go:233]: Instead of taking an generic `S` can this take []string like the parser's ParseStatements?

File: pkg/ottl/parser_collection.go
Comment (TylerHelmuth) [inline pkg/ottl/parser_collection.go:1]: I'm not against the use of reflection and generics (OTTL uses both a lot already), but we've also taken steps in the ottl context packages to make it easier to use OTTL in a component without needing to know about OTTL's generics.  `ottllog.NewParser` exists so users don't have to do `ottl.Parser[ottllog.TransformContext]`.  I believe we'll want to do something similar with this new API - It needs to be generic enough for others to provide their own context (thanks reflection), but our contexts should be able to hide it.

I believe this solution supports this but I want to call it out. In my head I want to still be able to do `ottllog.NewParser` and it takes care of using a `ParserCollection` behind the scenes, enforcing the valid contexts.

File: pkg/ottl/parser_collection.go (hunk 2 above)
Comment (edmocosta) [reply pkg/ottl/parser_collection.go:16]: Yes, it's an interface safe guard and is only being used to statically check the `ParseStatements` signature at compile time. It fails to compile If for some reason the `ottl.Parser.ParseStatements` changes (which could make the reflection [call](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36174/files/d2070b06eeb47b2860ad89939d808e38713a0da4#diff-7efe3affb096a0fa16edab4c8e48d665bd120c938103b778ad778b15c5f04ce6R47) to fail).

File: pkg/ottl/parser_collection.go (hunk 1 above)
Comment (edmocosta) [reply pkg/ottl/parser_collection.go:25]: Same as https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36174#discussion_r1829108905, but in this case it's not really necessary considering this interface is declared inside this same file. 
We can definitely replace it by a comment, but IMO, I'd keep it as a safe-guard, considering changing this function signature could break the parser collection [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36174/files/d2070b06eeb47b2860ad89939d808e38713a0da4#diff-7efe3affb096a0fa16edab4c8e48d665bd120c938103b778ad778b15c5f04ce6R77), and in all its external usages. WDYT?

File: pkg/ottl/parser_collection.go (hunk 3 above)
Comment (edmocosta) [reply pkg/ottl/parser_collection.go:38]: Good catch! that's correct, it does not need to be generic.

File: pkg/ottl/parser_collection.go (hunk 5 above)
Comment (edmocosta) [reply pkg/ottl/parser_collection.go:210]: That'd possible, but the idea of making it an interface was giving the`ParsedStatementConverter` functions more flexibility to receive extra data, without the need of adding an extra argument for some custom data.

For example, the `transformprocessor` builds the parser collection and them invokes the `ParseContextStatements` [function](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3c37b131871d76a21552b6b179d66a52831bba67/processor/transformprocessor/internal/common/logs.go#L107) passing on a custom type `ContextStatements` argument. Besides the `Statements []string` property, that type has a `Conditions []string` property used to convert ([see](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3c37b131871d76a21552b6b179d66a52831bba67/processor/transformprocessor/internal/common/logs.go#L71)) the parsed `ottl.Statement[K]` into `[R]`, which is only known after the parser collection is built.

In summary, having it as an interface allows the `ParsedStatementConverter` functions to access customised `ParseStatements` input arguments, avoiding extra parameters on the `ParsedStatementConverter` and making the parser collection reusable. 

WDYT?

File: pkg/ottl/parser_collection.go (hunk 4 above)
Comment (edmocosta) [reply pkg/ottl/parser_collection.go:110]: > Is S StatementsGetter required? 

Not really, I think we can drop it from the `ParserCollection`. It's only necessary for the `ParsedStatementConverter`.

File: pkg/ottl/parser_collection.go
Comment (edmocosta) [reply pkg/ottl/parser_collection.go:1]: > I believe this solution supports this but I want to call it out. In my head I want to still be able to do ottllog.NewParser and it takes care of using a ParserCollection behind the scenes, enforcing the valid contexts.

That's interesting, but I think it would require another type of API (builder perhaps) considering a single OTTL context (`ottllog.NewParser`) woudn't be able to build the whole parser collection. For example, if I want to parse OTTL `log` and `metric` contexts statements, I'd create a parser collection as:

```go
logParser, _ := ottllog.NewParser(ottlfuncs.StandardFuncs[ottllog.TransformContext](), telemetrySettings, ottllog.WithPathContextNames())
metricParser, _ := ottlmetric.NewParser(ottlfuncs.StandardFuncs[ottlmetric.TransformContext](), telemetrySettings, ottlmetric.WithPathContextNames())

pc, err := ottl.NewParserCollection[any](
  telemetrySettings,
  ottl.WithParserCollectionContext("log", &logParser, {...}),
  ottl.WithParserCollectionContext("metric", &metricsParser, {...}),
)

pc.ParseStatements(statements{[]string{`set(log.attributes["key"], "value")`}})
```
It indeed does not completely hides the generic type, but I think it's somehow similar to the `NewParser` functions.

>  In my head I want to still be able to do ottllog.NewParser and it takes care of using a ParserCollection behind the scenes, enforcing the valid contexts.

The valid contexts are still being enforced by the underline parser, and not by the parser collection, for example, given the previous  parser collection and the following `log` (new flat structure) statements:

```yml
- set(log.attributes["key"], "value")  # ok
- set(log.attributes["key"], attributes["value"]) # Parser fails and report a missing context for the `attributes["value"]` path
- set(log.attributes["key"], log.attributes["value"]) # ok
- set(log.attributes["key"], log.resource.attributes["value"]) # ok
- set(log.attributes["key"], log.instrumentation_scope.attributes["value"]) # ok
- set(log.attributes["key"], log.metric.attributes["value"]) # Parser fails as `metric` is not a valid path context name for the ottllog.NewParser (WithPathContextNames)
- set(trace.attributes["key"], trace.attributes["value"]) # ParserCollection fails as it does not support parsing `trace` context.
```

Please note that in this approach/initial version, the access to lower contexts would still be handled by the contexts themselves, for example [logs](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3c37b131871d76a21552b6b179d66a52831bba67/pkg/ottl/contexts/ottllog/log.go#L278), being the parser collection used only to parse top level contexts (inferred from the statements).

Comment (TylerHelmuth) [reply pkg/ottl/parser_collection.go:1]: >  For example, if I want to parse OTTL log and metric contexts statements, I'd create a parser collection as:

For now we can rely on `ottl.NewParserCollection` for this types of "mixed" scenarios, but for our standard situations, like wanting to parse log statements or span statements, the context can provide parsers that know how to parse only their kinds of statements.  The components like the transformprocessor or filterprocessor that work on multiple signals will need multiple parsers, but thats what they do now.

File: pkg/ottl/parser_collection.go (hunk 5 above)
Comment (TylerHelmuth) [reply pkg/ottl/parser_collection.go:210]: If it can at least be updated to use `StatementsGetter` as the argument instead of a generic that would be good.

File: pkg/ottl/parser_collection.go (hunk 4 above)
Comment (edmocosta) [reply pkg/ottl/parser_collection.go:110]: Giving it a second thought, if we drop it from `ParserCollection` and only use it on the `ParsedStatementConverter`, we wouldn't be able to enforce the `ParseStatements` argument to be the same type as the one used to add the context parser, for example: 

```go
type statementsA []string
func (s statementsA) GetStatements() []string { return s }

type statementsB []string
func (s statementsB) GetStatements() []string { return s }

func() {
  logParser, _ := ottllog.NewParser(ottlfuncs.StandardFuncs[ottllog.TransformContext](), telemetrySettings, ottllog.WithPathContextNames())

  pc, _ := ottl.NewParserCollection[any](
    componenttest.NewNopTelemetrySettings(),
    ottl.WithParserCollectionContext("log", &logParser, ottl.NewNopParsedStatementConverter[ottllog.TransformContext, statementsA]()),
  )

   // Panic as the expected type for the NewNopParsedStatementConverter function is statementsA, not statementsB
  pc.ParseStatements(statementsB([]string{`set(log.attributes["key"], log.resource.attributes["value"])`})) 
}
```

Another option would be dropping all `[S]` generic type usages replacing it by the `StatementsGetter` interface, delegating the type casting logic to the `ParsedStatementConverter` impls, example:

```go
ottl.NewParserCollection[any](
	telemetrySettings,
	ottl.WithParserCollectionContext("log", &logParser,
		func(
			collection *ottl.ParserCollection[any],
			parser *ottl.Parser[ottllog.TransformContext],
			context string,
			statements ottl.StatementsGetter, // it's currently typed statementsB
			parsedStatements []*ottl.Statement[ottllog.TransformContext]) (any, error) {
				
			b := statements.(statementsB)
                         ...
		}),
)
```

**Comment by edmocosta** [comment]
Hi @TylerHelmuth! I've replaced the generic type `S` by the `ottl.StatementsGetter` interface (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36174/commits/b63c5ac89b2613d4df9697710b946fa6a7b07c61) as suggested in a few comments, the only drawback I can think of was explained at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36174#discussion_r1829969255, which finally wouldn't be a big deal for consumers to handle it. 

In addition to that, I've also exposed a default `ottl.StatementsGetter`  [implementation](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36174/files#diff-7efe3affb096a0fa16edab4c8e48d665bd120c938103b778ad778b15c5f04ce6R37) so it does not required consumers to create their own type only to satisfy the interface.

Please let me know your thoughts.

File: pkg/ottl/parser_collection.go [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+	_ *ParserCollection[any],
+	_ *Parser[any],
+	_ string,
+	_ StatementsGetter,
+	_ []*Statement[any]) (any, error) {
+	return nil, nil
+}
+
+// StatementsGetter represents a set of statements to be parsed
+type StatementsGetter interface {
```
Comment (TylerHelmuth) [inline pkg/ottl/parser_collection.go:32]: For all the public API additions in this PR can you add a comment like:

`// Experimental: *NOTE* this API is subject to change or removal in the future.`

File: pkg/ottl/parser_collection.go [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+// parserCollectionParser holds an ottlParserWrapper and its respectively
+// statementsConverter function.
+type parserCollectionParser struct {
+	ottlParser          *ottlParserWrapper
+	statementsConverter statementsConverterWrapper
+}
+
+// ParserCollection is a configurable set of ottl.Parser that can handle multiple OTTL contexts
+// parsings, inferring the context and choosing the right parser for the given statements.
+type ParserCollection[R any] struct {
```
Comment (TylerHelmuth) [inline pkg/ottl/parser_collection.go:125]: In the godoc comments can you define what `R` represents.

**Review by TylerHelmuth — COMMENTED** [review]
@edmocosta I like the way this looks without the extra generic. I'd like to continue moving forward with this work because it will really help end users so I suggest we add `Experimental` to the new public API in this PR in cases we end up changing it later (which wouldn't affect end users, only component developers). We've used this strategy a bit in Core and while I don't want to leave the comment for a long time it will give us some coverage over the next couple releases.

File: pkg/ottl/parser_collection.go [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+	ottlParser          *ottlParserWrapper
+	statementsConverter statementsConverterWrapper
+}
+
+// ParserCollection is a configurable set of ottl.Parser that can handle multiple OTTL contexts
+// parsings, inferring the context, choosing the right parser for the given statements, and
+// transforming the parsed ottl.Statement[K] slice into a common result of type R.
+//
+// Experimental: *NOTE* this API is subject to change or removal in the future.
+type ParserCollection[R any] struct {
```
Comment (TylerHelmuth) [inline pkg/ottl/parser_collection.go:132]: Is there any type of restriction we can put on R? It doesn't feel like a `ottl.Statement[K] slice` could be turned into any other type. In the draft it looks like R is a `consumer.[Traces|Metrics|Logs]`, it would be nice to enforce R's type more if we can.

Comment (edmocosta) [reply pkg/ottl/parser_collection.go:132]: Not sure if we can easily implement it, I can give it a try, but I'm wondering which type restriction would fit here, considering the parser collection is dynamic and can be used to parse any type of context.

The advantage of having it as `any` is flexibility, as consumers could use any custom converter/aggregators on the converter functions, including using the raw `ottl.Statement[K] ` values and `switch r.(type) { case []*Statement[ottllog.TransformContext]: ... }` it later for processing, for example. 

If we don't have any major concern on not restricting it, I'd keep it as `any`, but I'm fine giving it a try if you think otherwise.  Thanks!

Comment (TylerHelmuth) [reply pkg/ottl/parser_collection.go:132]: It would be really great to do something to ensure `return convertedStatements.Interface().(R), nil` doesn't panic.  The `tCtx` (K) being an `any` ends up working out bc the type is only passed around, we never assert anything on it. But with R we attempt to do a cast.  If the `convertedStatements` can't be cast into R we'll panic right?

@evan-bradley what are your thoughts?

Comment (TylerHelmuth) [reply pkg/ottl/parser_collection.go:132]: Looking again at the draft implementation and how it hooks into the transformprocessor I see how this is enabling us to move the "Collection" idea from the processor into OTTL, which was our goal. I'm beginning to think it would be better, though, if we keep the Signal as part of the Collection, like the transformprocessor had it.  

We could potentially expose a generic solution in the future, but for the start I'm wondering if the API is safer if we expose a `MetricsParserCollection` that requires R to be a `consumer.Metrics` interface.

Comment (TylerHelmuth) [reply pkg/ottl/parser_collection.go:132]: We'd still have the panic opportunity in the happy-path return statment tho: `return convertedStatements.Interface().(R), nil`. There is no guarantee that `convertedStatements.Interface()` can be an R.

Comment (edmocosta) [reply pkg/ottl/parser_collection.go:132]: > We could potentially expose a generic solution in the future, but for the start I'm wondering if the API is safer if we expose a MetricsParserCollection that requires R to be a consumer.Metrics interface and the same for Traces and Logs.

IMO, In terms of safety, having strict types would definitely make it harder to consumers misusing the API, but in general, what works for `consumer.X` should also work for any type of `R`.

We would also need to include the `go.opentelemetry.io/collector/consumer` dependency into the ottl package. But it's true that testing would be easier considering we have a limited number of accepted types. 

> We'd still have the panic opportunity in the happy-path return statment tho: return convertedStatements.Interface().(R), nil. There is no guarantee that convertedStatements.Interface() can be an R.

I might be missing something, but it should be guaranteed (public API level) by the `WithParserCollectionContext` option (`ParserCollectionOption`), which requires both parser collection and converter to have the same R type.

Comment (TylerHelmuth) [reply pkg/ottl/parser_collection.go:132]: But does that guarantee that the `reflect.Value` returned from `statementsConverter.call` is of type R?

Comment (TylerHelmuth) [reply pkg/ottl/parser_collection.go:132]: > If I'm not missing anything, It should be enough to guarantee that all parser collection converters returns the same R type.

I agree with this.  I am more interested if, down in `ParseStatementsWithContext`  the value returned from `contextParser.statementsConverter.call` is guaranteed to be castable to `R`.

```go
        // Are we guaranteed that convertedStatements can be cast to an R?
        // I know the compiler will think we can cast this to R, but since it is 
        // a reflect.Value I don't think we have that guarantee. For example
        // if R is an int64, what happens here?
        convertedStatements, err := contextParser.statementsConverter.call(
		reflect.ValueOf(pc),
		contextParser.ottlParser,
		context,
		statements,
		parsedStatements,
	)

	if err != nil {
		return *new(R), err
	}

	return convertedStatements.Interface().(R), nil
```

Comment (TylerHelmuth) [reply pkg/ottl/parser_collection.go:132]: Oh wait I think I see it now. `ParsedStatementConverter` is a function that MUST return R. Thats the function that gets called here:

```go
contextParser.statementsConverter.call(
		reflect.ValueOf(pc),
		contextParser.ottlParser,
		context,
		statements,
		parsedStatements,
	)
```

So we're good. While it returns a reflect.Value, we KNOW that value is an R bc of the type of the `ParsedStatementConverter`.

Comment (TylerHelmuth) [reply pkg/ottl/parser_collection.go:132]: Assuming my understanding is correct, I am ok not introducing signal-specific stuff in the OTTL package. If needed we could add that to the ottlcontexts (if we wanted to hide the generics from the transformprocessor)

Comment (edmocosta) [reply pkg/ottl/parser_collection.go:132]: Yes, that's correct! It's guaranteed by the public API. 

I've pushed the extra condition to avoid panicking with `nil` return (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36174/commits/013c309eada994ec3a2210f6e21e102f9990ef7d)

**Review by TylerHelmuth — APPROVED** [review]

File: pkg/ottl/parser_collection_test.go [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+
+	pc, err := NewParserCollection(
+		component.TelemetrySettings{},
+		WithParserCollectionContext("foo", fooParser, failingConverter),
+		WithParserCollectionContext("bar", barParser, newNopParsedStatementConverter[any]()),
+	)
+	require.NoError(t, err)
+	pc.contextInferrer = &mockStaticContextInferrer{"bar"}
+
+	statements := mockStatementsGetter{values: []string{`set(bar.attributes["bar"], "foo")`, `set(bar.attributes["bar"], "bar")`}}
```
Comment (evan-bradley) [inline pkg/ottl/parser_collection_test.go:223]:
```suggestion
	// The `foo` context is never used, so these statements will successfully parse.
	statements := mockStatementsGetter{values: []string{`set(bar.attributes["bar"], "foo")`, `set(bar.attributes["bar"], "bar")`}}
```
Not strictly necessary, but I can't think of another way to explicitly capture the intent of this test (a failing parser that isn't called won't fail parsing).

File: pkg/ottl/parser_collection.go [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+	}
+}
+
+// ParseStatements parses the given statements into [R] using the configured context's ottl.Parser
+// and subsequently calling the ParsedStatementConverter function.
+// The statement's context is automatically inferred from the [Path.Context] values, choosing the
+// highest priority context found.
+// If no contexts are present in the statements, or if the inferred value is not supported by
+// the [ParserCollection], it returns an error.
+// If parsing the statements fails, it returns the underline [ottl.Parser.ParseStatements] error.
```
Comment (evan-bradley) [inline pkg/ottl/parser_collection.go:247]:
```suggestion
// If parsing the statements fails, it returns the underlying [ottl.Parser.ParseStatements] error.
```

**Review by evan-bradley — APPROVED** [review]
Thanks @edmocosta for continuing to work on this. Everything looks good to me, I only have wording suggestions.
